gocraft/work
# FEATURE: add MaxConcurrency control to job options

The purpose of this PR is to provide the `MaxConcurrency` job option that will control the max number of active jobs running at once across all workers for a specific job type (within one redis instance). For example:

  `worker_pool.JobWithOptions(jobName, JobOptions{MaxConcurrency: 1}, (*Context).WorkFxn)`

The above would ensure that jobs of that type never run more than 1 active job at a time. The default value is `MaxConcurrency = 0`, which is the current behavior (max concurrency would be the number of workers for that job * number of distinct processes running those workers).

This is accomplished by introducing a redis key that is used as a counting semaphore to lock a job queue. This lock is:

- checked before a new job is fetched
- incremented once a new job is fetched, if the jobs have `MaxConcurrency > 0`
- decremented once the job has finished (in the event of success or error)

We also check for any stale queue locks during worker pool startup and clean those up.

**QA tests run:**

1. Test Case 1: _verify only one job is active at a time_ 
    - create a worker pool with concurrency >= num of jobs and set `MaxConcurrency = 1` via JobOptions
    - start the worker pool
    - enqueue a bunch of jobs
    - during the duration of the job execution, check the `inprogress` queue and make sure there is never > 1 job in progress at any time
2. Test Case 2:  _new worker pool is starting up and detects an active queue lock, leaves lock in place_
    - manually create the queue lock and manually create an inprogress key
    - start a new worker pool in the same namespace/job name
    - verify that the stale lock is still in place
3. Test Case 3: _new worker pool is starting up and detects a stale queue lock_
    - manually create the queue lock and manually delete any inprogress keys
    - start a new worker pool in the same namespace/job name
    - verify that the stale lock was removed
4. Test Case 4: _verify that exclusive jobs can be paused_
    - start exclusive jobs
    - try to pause exclusive job execution 
    - observe that they stop running
    - unpause and observe that they resume execution
5. Functional tests for:
  - pausing / unpausing jobs
  - max concurrency default (no limit) and with limit
